src/main/scala/quasar/sbtdatasource/AssemblePlugin.scala

```diff
@@ -0,0 +1,205 @@
+      dsDependencies: Seq[ModuleID],
+      dsJar: Path,
+      quasarVersion: String,
+      dstDir: Path)(
+      implicit P: Parallel[F, G])
+      : F[Path] = {
+
+    // We assemble every component of the final tarball in this folder
+    val buildDir = dstDir.resolve("plugin-build")
+
+    // Will contain the dependencies and the plugin jar itself
+    val pluginDir = buildDir.resolve(dsName)
+
+    // The jar file's own path.
+    val jarPath = pluginDir.resolve(dsJar.getFileName)
+
+    // the datasource jar's path relative to the build dir
+    // included in the generated .plugin file to let quasar
+    // know where to load it from.
+    val relativeJarPath = buildDir.relativize(jarPath)
+
+    // the path to the generated .plugin file.
+    val pluginPath = buildDir.resolve(s"$dsName.plugin")
+
+    // start coursier on resolving all of the datasource's
+    // dependencies, *except* for quasar. quasar and its
+    // dependencies are already present in the user's
+    // `plugins` folder.
+    val resolution = Resolution(dsDependencies.map(moduleIdToDependency).toSet)
+
+    val cache = Cache.fetch[F](buildDir.toFile, CachePolicy.Update)
+
+    type OrFileErrors[A] = ValidatedNel[FileError, A]
+
+    for {
+      _ <- Sync[F] delay {
+        if (Files.exists(buildDir)) {
+          Files.walkFileTree(buildDir, RecursiveDeleteVisitor)
+        }
+      }
+
+      _ <- Sync[F].delay(Files.createDirectories(pluginDir))
+
+      _ <- Sync[F].delay(println("Fetching artifacts with coursier..."))
+
+      copyJar = Sync[F].delay(Files.copy(dsJar, jarPath, REPLACE_EXISTING))
+
+      fetchJarFiles = for {
+        metadata <- ResolutionProcess(resolution).run(
+          // we don't use ~/.ivy2/local here because
+          // I've heard that coursier doesn't copy
+          // from local caches into other local caches.
+          Fetch.from(Seq(MavenRepository("https://repo1.maven.org/maven2")), cache))
+
+        // fetch artifacts in parallel into cache
+        artifactsPar <-
+          metadata.artifacts.toList.parTraverse(f =>
+            Cache.file[F](f, pluginDir.toFile, CachePolicy.Update).run)
+
+        // some contortions to make sure *all* errors
+        // are reported when any fail to download.
+        artifacts <-
+          artifactsPar.traverse[OrFileErrors, File](_.toValidatedNel).fold(
+            es => Sync[F].raiseError(new Exception(s"Failed to fetch files: ${es.foldMap(e => e.toString + "\n\n")}")),
+            Sync[F].pure(_))
+
+        // filter out coursier metadata, we only want the jars
+        // for the `classpath` field of the .plugin file
+        jarFiles = artifacts.filter(_.getName.endsWith(".jar"))
+      } yield jarFiles
+
+      // coursier prefers that we fetch metadata before fetching
+      // artifacts. we do that in parallel with copying the datasource
+      // jar to its new place, because they don't depend on one
+      // another.
+      fetchedJarFiles <- copyJar &> fetchJarFiles
```

Comment:
I can sort of guess based on the comments, but I had never heard of `&>` before. I really hate all of the `Parallel` stuff. It's like, marginally better than scalaz's old `Nondeterminism`, but only marginally. Definitely not your fault, just observing.